### ISSUE TITLE
[RFC] Introducing the zero gradients optimization for AMP

### ISSUE THREAD

**Issue by ymwangg** [pytorch/xla#3120]
## Introduction
Based on our discussions in #3043, we propose a solution (PR: #3119) to remove the extra host-device sync for NaN checking in GradScaler. We saw ~20% speedup in the `test_train_mp_imagenet.py` example and some improvements in other models with this optimization. However, this method may be controversial as it changes the training algorithm. So, we'd like to get some feedbacks from the community.
The assumption of this optimization is that the following algorithm:
```python
if not has_nan:
    optimizer.step()
```
is equivalent to:
```python
if has_nan:
    optimizer.zero_grad()
optimizer.step()
```
Then we can use the following trick to perform the gradient update:
```python
found_inf = torch.stack(
     tuple(optimizer_state["found_inf_per_device"].values())).sum()
scaling_factor = 1.0 - found_inf.to(torch.bool).to(torch.float)
for grad in optimizer.gradients:
     grad.nan_to_num_()
     grad.mul_(scaling_factor)
optimizer.step()
```
This is a reasonable approximation as `has_nan=True` happens infrequently. We have tested several models and haven't found any counterexample that this method changes the training behavior.
## Caveats
1. One major concern is that this method changes the training algorithm and may pose unknown risks to users. We also have some debates on it.
2. We noticed increased memory consumption when using Adam optimizer investigated separately in #3102 due to the removal of the mark_step in GradScaler, though this issue is not specific to this optimization. One workaround is to add a barrier before calling `optimizer.step`.

PS: Special thanks to @codeislife99 for extensive testing/discussions and @trevor-m for helpful discussions.

CC: @JackCaoG @anijain2305 @byronyi

## Bechmark Results
Here are some benchmark results running on Nvidia Tesla V100 GPU.
To reproduce the numbers, run the following scripts in `xla/test` with different batch sizes:
```sh
GPU_NUM_DEVICES=1 python test_train_mp_mnist_amp.py --fake_data --batch_size 128
GPU_NUM_DEVICES=1 python test_train_mp_mnist_amp.py --fake_data --batch_size 128 --use_zero_grad
GPU_NUM_DEVICES=1 python test_train_mp_imagenet.py --fake_data --amp --batch_size 64
GPU_NUM_DEVICES=1 python test_train_mp_imagenet.py --fake_data --amp --batch_size 64 --use_zero_grad
```

| Model      | original | with zero_grad | speedup |
| ---------------------------| ----------- | ----------- | ----------- |
| Mnist (batch_size=128)       | 20061.19       |34958.32 | 1.74 |
| Mnist (batch_size=256)         | 32423.96        |56336.91 | 1.74 |
| Mnist (batch_size=512) | 51569.42 | 73367.12 | 1.42 |
| ImageNet (batch_size=64) | 733.75 | 981.75 | 1.34 |
| ImageNet (batch_size=128) | 987.57 | 1200.07 | 1.22 |
| ImageNet (batch_size=256) | 1186.94 | 1358.23 | 1.14 |

Note: The unit is samples / second.

**Comment by JackCaoG** [comment]
fyi @yaochengji

**Comment by mcarilli** [comment]
@cbcase and I figured out a recipe for syncfree dynamic gradient scaling over a year ago that works with eager execution, but it requires a custom optimizer with modified CUDA kernels.
The optimizer's CUDA kernels must take an additional `has_nan` parameter that's a pointer to the one-element found_inf _device_ tensor, then the beginning of each optimizer kernel checks this value _on the device_ and internally makes the decision to short circuit or run the math:
```c++
__global__ void syncfree_optimizer_kernel(..., found_inf_ptr) {
  if (*found_inf_ptr)
    return;
    
    optimizer math...
}
```
This way, the CPU Python code can enqueue inf/nan checking kernels that populate `found_inf` _on the device_, then `syncfree_optimizer_kernel` uses `found_inf` to make the right decision by checking this device value. In other words, the CPU can blindly enqueue everything (infcheck op followed by `optimizer.step()`) every iteration and trust that the kernels will do the right thing (either skip the step or run the math) with no CPU-side control flow decision needed.

This is how custom optimizers in our mlperf models work, but pytorch native optimizers certainly aren't written with the in-kernel short-circuit capability (for now), so they need a CPU-based control flow decision
```
if not found_inf_tensor.item():
    optimizer.step()
```

Your alternative CPU control flow
```
if has_nan:
    optimizer.zero_grad()
optimizer.step()
```
won't help avoid a sync, at least not for native optimizers with eager pytorch GradScaler. It still requires checking `has_nan` on the CPU, so you must retrieve it with `found_inf_tensor.item()` at some point.

I don't understand why you see a 20% speedup, but I'd really like to. Can you explain your approach in more detail? Does XLA do some magic here I don't understand?

**Comment by byronyi** [comment]
The pseudo code should be changed to 

```
# summing device tensors still yields a device tensor, so no sync involved
found_inf = torch.stack(optimizer_state["found_inf_per_device"].values()).sum()
for grad in optimizer.gradients:
  grad.mul_(1.0 - found_inf.cast(torch.boolean)))
```

And then the XLA could (not necessarily though) fuse these operations into a custom CUDA kernel, similar to `syncfree_optimizer_kernel`.

**Comment by mcarilli** [comment]
ohh gotcha, that'll work. But, as @ymwangg says, it's still true that for some optimizers (e.g. optimizers with momentum)
```
optimizer.zero_grad()
optimizer.step()
```
is not numerically equivalent to skipping `optimizer.step()`.

**Comment by byronyi** [comment]
Back quoting @mcarilli's comment in https://github.com/pytorch/pytorch/issues/61880#issuecomment-917300300

> You could also consider graph-capturing optimizer.step() by itself, then patching optimizer.step with a one-line function that runs the graph. Later, when you run scaler.step(optimizer) ungraphed in the training loop, the inf/nan check will take place ungraphed, but when scaler.step(optimizer) internally chooses to run optimizer.step(), your patched optimizer.step will run the optimizer kernels as a graph.

I am wondering if we could unify the Python API in here for both CUDA graph and torch-xla (or Lazy Tensor in core), so the users could supply additional hint on which part of the graph-capture are intended for reuse, e.g. wrap `model.forward()` or `optimizer.step()` in torch.fx. This hint could potentially reduce the re-tracing overhead while still keeping the great usability of lazy tensor even without the hint.

cc @wconstab @ezyang

**Comment by ymwangg** [comment]
@byronyi @mcarilli Thanks for bringing up the idea of `sync_free_optimizer`. We think another good alternative is to implement the `sync_free_optimizer` using XLA builder inside PT_XLA and it does not change the training algorithm.

**Comment by JackCaoG** [comment]
@ymwangg I think using [`conditional`](https://github.com/pytorch/xla/blob/08ae1044c2a7e314895f9946104cbe399e096515/torch_xla/core/xla_builder.py#L526) in the op builder would be a way to go. However I think we are lacking an easy way to convert a python function(optimzier) into a `XlaComputation` which is the input to the `conditional`. I image you can also handcraft `XLAComputation` in this case but it might takes longer.

[functorch](https://github.com/facebookresearch/functorch) that @zou3519 and his team is working on might help in this case (although I don' think it is the intent of the functorch project 😄 ); after pytorch/xla actually support functorch.

**Comment by ezyang** [comment]
Everything discussed here seems reasonable. This being eager mode we'll probably need a custom kernel for it, but it definitely seems worth doing.

**Comment by wconstab** [comment]
> I am wondering if we could unify the Python API in here for both CUDA graph and torch-xla (or Lazy Tensor in core),
Yes, we have not designed this yet for lazy tensor core, but we are thinking about having some explicit API to avoid retracing in known-static regions.  I hadn't thought about aligning the API with the cuda-graph one, but maybe that could work.
